packages/category-exporter/src/main.js

```diff
@@ -0,0 +1,97 @@
+/* @flow */
+import { createClient } from '@commercetools/sdk-client'
+import { createRequestBuilder } from '@commercetools/api-request-builder'
+import { createHttpMiddleware } from '@commercetools/sdk-middleware-http'
+import {
+  createAuthMiddlewareForClientCredentialsFlow,
+  createAuthMiddlewareWithExistingToken,
+} from '@commercetools/sdk-middleware-auth'
+import { createUserAgentMiddleware } from '@commercetools/sdk-middleware-user-agent'
+import fetch from 'node-fetch'
+import type {
+  ApiConfigOptions,
+  CategoryExporterOptions,
+  LoggerOptions,
+} from 'types/categoryExporter'
+import type { Client } from 'types/sdk'
+import silentLogger from './utils/silent-logger'
+import pkg from '../package.json'
+
+export default class CategoryExporter {
+  // Set type annotations
+  apiConfig: ApiConfigOptions
+  client: Client
+  logger: LoggerOptions
+  predicate: ?string
+
+  constructor(options: CategoryExporterOptions) {
+    if (!options.apiConfig)
+      throw new Error('The constructor must be passed an `apiConfig` object')
+    this.apiConfig = options.apiConfig
+    this.client = createClient({
+      middlewares: [
+        createAuthMiddlewareWithExistingToken(
+          options.accessToken ? `Bearer ${options.accessToken}` : ''
+        ),
+        createAuthMiddlewareForClientCredentialsFlow({
+          ...this.apiConfig,
+          fetch,
+        }),
+        createUserAgentMiddleware({
+          libraryName: pkg.name,
+          libraryVersion: pkg.version,
+        }),
+        createHttpMiddleware({
+          host: this.apiConfig.apiUrl,
+          fetch,
+        }),
+      ],
+    })
+
+    this.predicate = options.predicate
+
+    this.logger = {
+      ...silentLogger,
+      ...options.logger,
+    }
+  }
+
+  run(outputStream: stream$Writable): Promise<any> {
+    this.logger.info('Starting Category Export')
+    return this.client
+      .process(
+        { uri: this.buildURI(), method: 'GET' },
+        (payload: Object): Promise<any> => {
+          if (payload.statusCode !== 200)
+            return Promise.reject(
+              new Error(`Request returned error ${payload.statusCode}`)
+            )
+          const results = payload.body.results
+          const JSONResults = JSON.stringify(results)
+          outputStream.write(JSONResults)
```

Comment:
This solution will fail if we try to export a lot of categories as it will write each array individually. It will look something like below

```
[
  ...categoryObjects
]
[
  ...categoryObjects
]
[
  ...categoryObjects
]
```

One solution would be to simply destruct all of the arrays and write all objects into a single one instead. Another one would be to pipe them all through something like [JSONStream](https://github.com/dominictarr/JSONStream).